airflow/decorators/base.py

```diff
@@ -208,11 +208,36 @@ def __init__(
         # since values for those will be provided when the task is run. Since
         # we're not actually running the function, None is good enough here.
         signature = inspect.signature(python_callable)
+
+        # Don't allow context argument defaults other than None to avoid ambiguities.
+        if any(
+            param.name in KNOWN_CONTEXT_KEYS and param.default not in (None, inspect.Parameter.empty)
+            for param in signature.parameters.values()
+        ):
+            raise ValueError("Context key parameters can't have a default other than None")
+
         parameters = [
             param.replace(default=None) if param.name in KNOWN_CONTEXT_KEYS else param
             for param in signature.parameters.values()
         ]
-        signature = signature.replace(parameters=parameters)
+        try:
+            signature = signature.replace(parameters=parameters)
+        except ValueError as err:
+            raise ValueError(
+                textwrap.dedent(
+                    f"""
+                    The function signature broke while assigning defaults to context key parameters.
+
+                    The decorator is replacing the signature
+                    > {python_callable.__name__}({', '.join(str(param) for param in signature.parameters.values())})
+
+                    with
+                    > {python_callable.__name__}({', '.join(str(param) for param in parameters)})
+
+                    which isn't valid: {err}
+                    """
+                )
```

Comment:
Smal nit. I would rather put this into the variable and use it as exception message, it improve readability of the traceback.

eg. 
```python
msg = textwrap.dedent
    ...
)
raise ValueError(msg)

Reply:
Traceback looks much better with this change, thanks for the suggestion